# Name: Neelanjan Dutta

# Register number: 2448040

# Program #1: Implementing a Simple Perceptron for Real-Time Binary Classification.

### (a)Select a real-time scenario where a binary classification is applicable. Some examples include(Classifying emails as spam or not spam,Predicting whether a patient has diabetes or not,Identifying if a customer will churn or stay,Classifying whether a stock market trend is up or down,Classifying a person as employed or unemployed).

### Answer:

<b> Selected Scenario: </b>
Classifying whether a patient has heart disease based on clinical and physiological features.<br>
This is a real-world binary classification task where: <br>
(i) <b>Presence (1)</b> means the patient is diagnosed with heart disease. <br>
(ii) <b>Absence (0)</b> means the patient does not have heart disease.<br>

### (b)Choose a dataset relevant to your scenario. You may Use publicly available datasets (e.g., UCI Machine Learning Repository, Kaggle.Generate a simple synthetic dataset if needed.

### Answer:

<b>Selected Dataset:</b> <br>

Heart Disease UCI Dataset<br>
<br>
<b>Dataset Source:</b> <br>

Available from the UCI Machine Learning Repository<br>

Preprocessed and commonly used version available on Kaggle<br>

### c) Implement the Perceptron algorithm from scratch,including:Initialization of weights and bias,Weight update rule using gradient descent,Activation function (step function or threshold function),Training loop based on epochs.

### Answer:

In [6]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the dataset 
df = pd.read_csv("C:/Users/Neelanjan Dutta/Downloads/heart.csv")

# Features and target
X = df.drop('target', axis=1)  # 13 features
y = df['target']               # 1 = Heart disease, 0 = No disease

# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the dataset into training and testing sets (80-20 split)
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

In [7]:
# Define Perceptron Model from Scratch
class Perceptron:
    def __init__(self, input_dim, lr=0.01, epochs=100):
        self.lr = lr
        self.epochs = epochs
        self.weights = np.zeros(input_dim)
        self.bias = 0

    def activation(self, x):
        # Step/Threshold Activation Function
        return 1 if x >= 0 else 0

    def predict(self, x):
        z = np.dot(x, self.weights) + self.bias
        return self.activation(z)

    def fit(self, X, y):
        best_accuracy = 0
        patience = 5     # Number of epochs to wait if no improvement
        wait = 0
        accuracy_list = []

        for epoch in range(self.epochs):
            total_error = 0
            for xi, target in zip(X, y):
                z = np.dot(xi, self.weights) + self.bias
                y_pred = self.activation(z)
                error = target - y_pred

                # Weight update rule using gradient descent
                self.weights += self.lr * error * xi
                self.bias += self.lr * error
                total_error += abs(error)

            # Calculate training accuracy
            predictions = [self.predict(xi) for xi in X]
            accuracy = np.mean(predictions == y)
            accuracy_list.append(accuracy)

            print(f"Epoch {epoch+1}, Total Errors: {total_error}, Training Accuracy: {accuracy:.4f}")

            # Smart Early Stopping: stop if accuracy doesn't improve
            if accuracy > best_accuracy:
                best_accuracy = accuracy
                wait = 0
            else:
                wait += 1
                if wait >= patience:
                    print(f"Stopping early. Accuracy did not improve for {patience} consecutive epochs.")
                    break

# Create the model and train
model = Perceptron(input_dim=X_train.shape[1], lr=0.01, epochs=100)
model.fit(X_train, y_train)

# Evaluate on test data
y_pred = [model.predict(x) for x in X_test]
test_accuracy = np.mean(y_pred == y_test)
print(f"\nTest Accuracy: {test_accuracy:.4f}")

Epoch 1, Total Errors: 156, Training Accuracy: 0.7585
Epoch 2, Total Errors: 164, Training Accuracy: 0.7561
Epoch 3, Total Errors: 169, Training Accuracy: 0.7476
Epoch 4, Total Errors: 163, Training Accuracy: 0.7524
Epoch 5, Total Errors: 152, Training Accuracy: 0.7573
Epoch 6, Total Errors: 157, Training Accuracy: 0.7805
Epoch 7, Total Errors: 164, Training Accuracy: 0.7780
Epoch 8, Total Errors: 170, Training Accuracy: 0.7671
Epoch 9, Total Errors: 159, Training Accuracy: 0.7720
Epoch 10, Total Errors: 164, Training Accuracy: 0.7768
Epoch 11, Total Errors: 154, Training Accuracy: 0.7427
Stopping early. Accuracy did not improve for 5 consecutive epochs.

Test Accuracy: 0.6927


A Perceptron model is implemented from scratch using Python and NumPy.<br>
<br>
Weights and bias are initialized to zero.<br>
<br>
A step (threshold) activation function is used: outputs 1 if the input ≥ 0, else 0. <br>
<br>
For each training sample, the model computes the weighted sum and applies the activation function to make a prediction.<br>
<br>
The prediction is compared to the actual label, and the weights and bias are updated using the Perceptron learning rule based on the prediction error.<br>
<br>
Total error per epoch is tracked to monitor convergence, and training accuracy is computed after each epoch.<br>
<br>
Early stopping is applied: training halts if the accuracy does not improve for 5 consecutive epochs.<br>
<br>
Finally, the trained model is evaluated on unseen test data, and the test accuracy is reported.<br>

# Program #2: Implementing an ADALINE Neural Network for AND Logic Gate and Real-Time Binary Classification.

### a)Implement the ADALINE model using Python with the following features:

In [10]:
import numpy as np

# Bipolar AND gate
X = np.array([[1, 1], [1, -1], [-1, 1], [-1, -1]])
T = np.array([1, -1, -1, -1])

# Initialization
w = np.zeros(2)
b = 0.0
alpha = 0.1
max_epochs = 100
mse_tolerance = 0.0001  # how small an improvement we expect
prev_mse = float('inf')

for epoch in range(max_epochs):
    print(f"\nEpoch {epoch + 1}")
    total_error = 0

    for i in range(len(X)):
        x = X[i]
        t = T[i]

        y = np.dot(w, x) + b
        error = t - y

        # Update weights and bias
        w += alpha * error * x
        b += alpha * error
        total_error += error ** 2

        print(f"Input: {x}, Target: {t}, Output: {y:.2f}, Error: {error:.2f}, Weights: {np.round(w, 3)}, Bias: {b:.3f}")

    # After epoch: compute overall predictions and MSE
    outputs = np.dot(X, w) + b
    predictions = np.sign(outputs)
    mse = total_error / len(X)

    print(f"MSE = {mse:.4f}, Outputs = {np.round(outputs, 2)}, Predictions = {predictions}")

    if not np.array_equal(predictions, T):
        print(f"\nStopped at epoch {epoch + 1}: Predictions no longer match targets.")
        break

    if prev_mse - mse < mse_tolerance:
        print(f"\nStopped at epoch {epoch + 1}: MSE no longer improving significantly (Δ={prev_mse - mse:.6f}).")
        break

    prev_mse = mse

# Final results
print("\nFinal Weights:", w)
print("Final Bias:", b)
print("Final Outputs:", np.round(outputs, 2))
print("Final Predictions:", predictions)


Epoch 1
Input: [1 1], Target: 1, Output: 0.00, Error: 1.00, Weights: [0.1 0.1], Bias: 0.100
Input: [ 1 -1], Target: -1, Output: 0.10, Error: -1.10, Weights: [-0.01  0.21], Bias: -0.010
Input: [-1  1], Target: -1, Output: 0.21, Error: -1.21, Weights: [0.111 0.089], Bias: -0.131
Input: [-1 -1], Target: -1, Output: -0.33, Error: -0.67, Weights: [0.178 0.156], Bias: -0.198
MSE = 1.0304, Outputs = [ 0.14 -0.18 -0.22 -0.53], Predictions = [ 1. -1. -1. -1.]

Epoch 2
Input: [1 1], Target: 1, Output: 0.14, Error: 0.86, Weights: [0.264 0.242], Bias: -0.111
Input: [ 1 -1], Target: -1, Output: -0.09, Error: -0.91, Weights: [0.173 0.333], Bias: -0.203
Input: [-1  1], Target: -1, Output: -0.04, Error: -0.96, Weights: [0.269 0.238], Bias: -0.298
Input: [-1 -1], Target: -1, Output: -0.80, Error: -0.20, Weights: [0.289 0.257], Bias: -0.318
MSE = 0.6327, Outputs = [ 0.23 -0.29 -0.35 -0.86], Predictions = [ 1. -1. -1. -1.]

Epoch 3
Input: [1 1], Target: 1, Output: 0.23, Error: 0.77, Weights: [0.366 0.33

- This program implements the ADALINE (Adaptive Linear Neuron) model for a bipolar AND logic gate. <br>
- Inputs and targets are given in bipolar format, Inputs: [1, 1], [1, -1], [-1, 1], [-1, -1] Targets: [1, -1, -1, -1]<br>
- Initial weights and bias are set to zero. A linear output (raw net input) is used during learning instead of a threshold.<br>
- The original learning rate (α = 1) was reduced to α = 0.1 for the following reason:<br>
<br>
ADALINE uses raw (continuous) outputs to minimize Mean Squared Error (MSE).<br>
A high learning rate like α = 1 causes large weight updates, leading to unstable learning and exploding error.<br>
Reducing the learning rate to 0.1 ensures gradual, stable convergence of the model.<br>
<br>
<b>For each epoch:</b><br>
The model computes output y = w·x + b . Calculates error t − y <br>
Updates weights and bias using the delta rule (gradient descent). The model monitors Mean Squared Error (MSE) after every epoch.<br>
<br>
<b>Training stops early if:</b><br>
- All predictions match the target values (perfect learning), or <br>
- Change in MSE is very small (< 0.0001), indicating convergence<br>
<br>
Final learned weights, bias, outputs, and predictions are printed.<br>

### b)Real-Time Binary Classification with ADALINE:Choose a real-world binary classification scenario (ex.)Predict if a patient has diabetes (Yes/No),Predict customer churn (Churn/No Churn) etc.Select or load an appropriate dataset,Train an ADALINE model on the data.

### Answer

In [11]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the Heart Disease dataset
df = pd.read_csv("C:/Users/Neelanjan Dutta/Downloads/heart.csv")

X = df[['age', 'trestbps', 'chol']] 

# Target: 1 = heart disease, 0 = no disease
y = df['target']

# Convert labels to bipolar format: 0 → -1, 1 → +1
y = np.where(y == 0, -1, 1)

# Standardize features (important for ADALINE convergence)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split 
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

<b>Dataset Used:</b> The Heart Disease Dataset (commonly from UCI or Kaggle) is used for binary classification of patients: (0 → No Disease, 1 → Presence of Disease) <br>
<b>Classification Task:</b> The goal is to train an ADALINE model to classify whether a person has heart disease or not, based on clinical measurements.<br>
<b>Feature Selection (Only 3 Features):</b> We selected only 3 numerical features: age, resting blood pressure (trestbps), and cholesterol (chol).<br>
<br>
This is done to:<br>
Keep the model simple and interpretable<br>
Reduce computational complexity and help visualize weight updates<br>
Ensure better stability for ADALINE, which is sensitive to dimensionality<br>
These 3 features are also known to be medically relevant in predicting heart disease.<br>
<br>
Target Conversion to Bipolar Form:<br>
ADALINE works with bipolar outputs. So, the binary targets are converted as:<br>
(0 → -1 (no heart disease) 1 → +1 (presence of heart disease))<br>

In [12]:
import numpy as np

# Define ADALINE Model
class ADALINE:
    def __init__(self, input_dim, lr=0.01, epochs=100):
        self.lr = lr
        self.epochs = epochs
        self.weights = np.zeros(input_dim)
        self.bias = 0.0  # Ensure float

    def net_input(self, x):
        return np.dot(x, self.weights) + self.bias

    def activation(self, x):
        return x  # Linear activation

    def predict(self, X):
        output = self.net_input(X)
        return np.where(output >= 0, 1, -1)

    def fit(self, X, y):
        prev_mse = float('inf')
        mse_tolerance = 0.0001

        for epoch in range(self.epochs):
            total_error = 0
            for xi, target in zip(X, y):
                output = self.net_input(xi)
                error = target - output

                # ADALINE weight update rule
                self.weights += self.lr * error * xi
                self.bias += self.lr * error

                total_error += error ** 2

            mse = total_error / len(X)
            predictions = self.predict(X)
            accuracy = np.mean(predictions == y)

            print(f"Epoch {epoch+1} | MSE: {mse:.4f} | Accuracy: {accuracy:.4f}")

            # Early Stopping
            if np.array_equal(predictions, y):
                print(f"\nStopped early at epoch {epoch+1}: Perfect classification on training set.")
                break
            if abs(prev_mse - mse) < mse_tolerance:
                print(f"\nStopped early at epoch {epoch+1}: MSE stabilized (Δ={abs(prev_mse - mse):.6f})")
                break

            prev_mse = mse

# Train the model on Heart Disease dataset
model = ADALINE(input_dim=X_train.shape[1], lr=0.01, epochs=100)
model.fit(X_train, y_train)

# Evaluate on test data
y_pred = model.predict(X_test)
test_accuracy = np.mean(y_pred == y_test)
print(f"\nFinal Test Accuracy: {test_accuracy:.4f}")

Epoch 1 | MSE: 0.9616 | Accuracy: 0.6427
Epoch 2 | MSE: 0.9615 | Accuracy: 0.6427

Stopped early at epoch 2: MSE stabilized (Δ=0.000065)

Final Test Accuracy: 0.6000


<b>Model Used:</b> An ADALINE (Adaptive Linear Neuron) model is defined and trained from scratch.<br>
Training Method: The model uses:<br>
Linear activation (output = w·x + b) during training<br>
Sign function (+1 / -1) for binary prediction<br>
<b>Learning Rate:</b><br>
lr = 0.01 is chosen to ensure stable and gradual learning via gradient descent.<br>
<b>Loss Function:</b><br>
The model minimizes Mean Squared Error (MSE) between predicted continuous outputs and target labels.<br>
<b>Weight Updates:</b><br>
Weights and bias are updated after each training sample using the delta learning rule:<br>
<center>w = w + α⋅(target − output)⋅x</center> <br>
<center>b = b + α⋅(target − output)</center><br>
<b>Stopping Criteria Used:</b><br>
If MSE stops improving significantly (Δ < 0.0001)<br>
If all training predictions match the targets perfectly<br>
<b>Output Observation:</b>
Epoch 1 | MSE: 0.9616 | Accuracy: 0.6427<br>
Epoch 2 | MSE: 0.9615 | Accuracy: 0.6427<br>
Stopped early at epoch 2: MSE stabilized (Δ=0.000065)<br>
Final test accuracy on unseen data was 60.00%  <br>
<br>
<b>Interpretation:</b><br>
The ADALINE model converged quickly due to standardized features and a conservative learning rate.<br>
However, test accuracy was limited (60%) because the Heart Disease dataset is not linearly separable, and ADALINE is a linear model.<br>
Using only 3 features (age, trestbps, chol) further restricts the model’s ability to capture complex decision boundaries.<br>
Despite lower accuracy, the model training behaved as expected, demonstrating stable convergence and early stopping.<br>